Fix incorrect path in error overlay on Win

fcFn opened the pull request:
## Changes

Fix #6678.

## Testing

An e2e test was added.
<!-- How was this change tested? -->
<!-- DON'T DELETE THIS SECTION! If no tests added, explain why. -->

## Docs
This just fixes a bug so no docs changes are needed.
<!-- Could this affect a user’s behavior? We probably need to update docs! -->
<!-- If docs will be needed or you’re not sure, uncomment the next line: -->
<!-- /cc @withastro/maintainers-docs for feedback! -->

<!-- DON'T DELETE THIS SECTION! If no docs added, explain why.-->
<!-- https://github.com/withastro/docs -->

fcFn commented:
Ah, it seems that the title in the file path that is shown in the error overlay is supposed to be absolute while the text is supposed to be relative to the project root? That's how it is on Linux. 

Well, I fixed the double issue path on Windows so that now it at least shows a valid Windows path and the `Open in editor` button works on Windows (and the tests now pass on both Linux and Windows). 

Should I work some more on this and make it show the relative path in the text content like it does on Linux?

fcFn commented:
I just discovered that while fixing the path on Windows when the error is in `pages` the path will now break if the error is in a component 😅 I will investigate further.

fcFn commented:
I fixed it and added another test to check the path when an error is in a component.

Princesseuh commented on packages/astro/src/core/errors/dev/utils.ts:
```diff
@@ -32,10 +32,15 @@ export function collectErrorMetadata(e: any, rootFolder?: URL | undefined): Erro
 			error.loc?.file &&
 			rootFolder &&
-			(!error.loc.file.startsWith(rootFolder.pathname) || !isAbsolute(error.loc.file))
+			((!error.loc.file.startsWith(rootFolder.pathname) && !isWindows) ||
```
The path utilities in `packages/astro/src/core/path.ts` might help you make this a bit cleaner. An explanation of what it is doing would also be great!

fcFn commented on packages/astro/src/core/errors/overlay.ts:
```diff
@@ -636,7 +636,8 @@ class ErrorOverlay extends HTMLElement {
 
 			if (codeHeader) {
-				const cleanFile = err.loc.file.split('/').slice(-2).join('/');
+				const separator = err.loc.file.includes('/') ? '/' : '\\';
```
Overlay didn't account for Windows separators here which and this change fixes it.

fcFn commented on packages/astro/src/core/errors/dev/utils.ts:
```diff
@@ -32,10 +33,15 @@ export function collectErrorMetadata(e: any, rootFolder?: URL | undefined): Erro
+
+		// Normalize the paths so that we can correctly detect if it's absolute on any platform
+		const normalizedFile = error.loc?.file?.replace(/\\/g, '/');
+		const normalizedRootFolder = removeLeadingForwardSlashWindows(rootFolder?.pathname || '');
```
Remove a leading forward slash which `rootFolder.pathname` comes with which is invalid on Windows (e.d. `/c:/`) and will always fail the check for the path being absolute below which results in the bug that is described in the issue.

fcFn commented on packages/astro/src/core/errors/dev/utils.ts:
```diff
@@ -32,10 +33,15 @@ export function collectErrorMetadata(e: any, rootFolder?: URL | undefined): Erro
 		// - The "Open in Editor" button won't work
+
+		// Normalize the paths so that we can correctly detect if it's absolute on any platform
+		const normalizedFile = error.loc?.file?.replace(/\\/g, '/');
```
Normalize the separators because if we are on Windows the path sometimes uses `/` (if the error is in a component) and sometimes uses `\\` (if the error is in a page).

fcFn replied:
@Princesseuh Thank you so much for pointing that out! Pushed a fix.

fcFn commented on packages/astro/e2e/errors.test.js:
```diff
@@ -52,6 +52,28 @@ test.describe('Error display', () => {
 		expect(await page.locator('vite-error-overlay').count()).toEqual(0);
 	});
 
+	test('shows correct file path when a page has an error', async ({ page, astro }) => {
```
Two different tests because on Windows the path separators are different depending on the location of the error (i.e. component vs page).

fcFn commented:
@Princesseuh Thank you for the hint about the path utilities! 

I think I found the underlying cause and made the fix more robust. I added a comment for the less clear part and I added more explanations in the comments for the PR as I'm not sure which of those should be added to the code.
